nemo/core/neural_modules.py

```diff
@@ -113,40 +113,35 @@ def __extract_init_params(self) -> Dict[str, Any]:
             Dictionary containing parameters passed to init().
         """
         # Get names of arguments of the original module init method.
-        init_keys = getfullargspec(type(self).__init__).args
+        to_set_params = getfullargspec(type(self).__init__).args
+        to_set_params.remove("self")
 
-        # Remove self.
-        if "self" in init_keys:
-            init_keys.remove("self")
+        # Create empty list of init params.
+        init_params = {}
 
-        # Create a list of params and initialize it with a special value.
-        init_params = {}.fromkeys(init_keys, "__UNSET__")
-        no_of_unset = len(init_params)
-
-        # Retrieve values of those params from the call list.
-        # Do it by removing and analysing the calls from stack one by one.
+        # Get the frame "call context".
         for frame in stack()[1:]:
-            # Get call "context".
-            localvars = getargvalues(frame[0]).locals
-            # Check if we are in the "context" of the class call.
-            if "__class__" not in localvars.keys():
-                continue
-            # Check if this is the context of the current "class".
-            if type(localvars["self"]).__name__ != localvars["__class__"].__name__:
-                # If own class is not equal to the call context class.
-                continue
-            # Ok, got the actual __init__() call!!
-            # Copy the keys.
-            for key in init_keys:
-                # Found the variable - and it is still unset!
-                if key in localvars.keys() and init_params[key] == "__UNSET__":
-                    # Save the value.
-                    init_params[key] = localvars[key]
-                    no_of_unset -= 1
-            # That should set all the init_params!
-            assert no_of_unset == 0
-            # Ok, we can terminate.
-            break
+            # Get the call arguments.
+            localvars = getargvalues(frame[0])
+
+            # Fill the parameters with call_args.
+            for key in to_set_params:
+                if key in localvars.args:
+                    init_params[key] = localvars.locals[key]
+
+            # Remove all set keys.
+            for key in init_params.keys():
+                if key in to_set_params:
+                    to_set_params.remove(key)
+
+            # Check if we have set everything.
+            if len(to_set_params) == 0:
+                break
+
+        # Make sure that we collected ALL (and ONLY) the signature params - if not, then there is a BUG!
+        assert len(to_set_params) == 0
```

Comment:
Please do not use "assert" - raise appropriate exception instead with descriptive message.

Reply:
Ok, I already fixed that in a separate PR:
https://github.com/NVIDIA/NeMo/pull/639/files#diff-54953aa9df0c54722a7d1cf5c66de97aR141
that was merged with master:
https://github.com/NVIDIA/NeMo/blob/99ef493e8ea3a84324743dc76157b59d9bac6c22/nemo/core/neural_modules.py#L143
I simply haven't updated this branch yet...

Reply:
Merged with master. Plus polished the "error message" even more.

@okuchaiev Can we merge this PR? Thanks.